### PR TITLE
Added match query support. First stab

### LINKED ISSUE #3285: Add support for ES "match" requests.

### PR DESCRIPTION
**Pull request by fulmicoton**
Closes #3285

This PR also does some refactoring, to
extend the usage of the PhraseQuery (now FullTextQuery)
to cover the MatchQuery as well.

# How is tested?

Unit test.
I'll add integration tests once #3315 is merged.

### CODE REVIEW

File: quickwit/quickwit-query/src/elastic_query_dsl/match_query.rs [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+    #[serde(default)]
+    default_operator: BooleanOperand,
+    #[serde(default)]
+    zero_terms_query: MatchAllOrNone,
+}
+
+impl ConvertableToQueryAst for MatchQuery {
+    fn convert_to_query_ast(self) -> anyhow::Result<QueryAst> {
+        // TODO fix this.
+        // This is not exactly what we want. This will interpret boolean operators for instance.
```
Comment (trinity-1686a) [inline quickwit/quickwit-query/src/elastic_query_dsl/match_query.rs:54]: I don't think this is true?

Comment (fulmicoton) [reply quickwit/quickwit-query/src/elastic_query_dsl/match_query.rs:54]: ah yes this is an old comment.

Comment (fulmicoton) [inline quickwit/quickwit-query/src/elastic_query_dsl/match_query.rs:54]:
```suggestion
```

File: quickwit/quickwit-query/src/query_ast/term_set_query.rs [hunk 2]
```diff
@@ -37,7 +41,11 @@ impl TermSetQuery {
         for (full_path, values) in &self.terms_per_field {
             for value in values {
                 // We simplify the logic by calling compute_query, and extract the resulting terms.
```
[the remark this answers is not in the document]
Comment (fulmicoton) [reply quickwit/quickwit-query/src/query_ast/term_set_query.rs]: I fixed the comment.

File: quickwit/quickwit-query/src/query_ast/tantivy_query_ast.rs [hunk 3]
```diff
@@ -144,12 +180,40 @@ impl TantivyBoolQuery {
...
+            MatchAllOrNone::MatchNone,
+            has_no_positive_ast_so_far,
+        );
+        if self.must.is_empty()
+            && self.should.is_empty()
+            && self.must_not.is_empty()
+            && self.filter.is_empty()
+        {
+            return TantivyQueryAst::match_none();
+        }
```
Comment (trinity-1686a) [inline quickwit/quickwit-query/src/query_ast/tantivy_query_ast.rs:216]: I have a hard time wrapping my head around this bit.
1st I don't think I understand why the same condition above should `match_all`, but this one, after more filtering, should `match_none`.
2nd I'm not convinced the condition can ever evaluate to `true`. If any of the 4 lists is not empty before, due to the cascading `stop_before_empty = has_no_positive_ast_so_far`, then at least one of the list shouldn't be empty afterward?

Comment (fulmicoton) [reply quickwit/quickwit-query/src/query_ast/tantivy_query_ast.rs]: you are correct!

**Comment by fulmicoton** [comment]
Merging to allow pursuing work on #3296. Can you have a look at the last few commits on Monday @trinity-1686a ?

**Comment by trinity-1686a** [comment]
looks good

**Pull request merged**